src/jaxsim/integrators/common.py

```diff
@@ -0,0 +1,416 @@
+        valid = valid and A.shape == (c.size, b.T.shape[1])
+        valid = valid and bool(jnp.all(b.T.sum(axis=1) == 1))
+
+        return valid
+
+    @staticmethod
+    def butcher_tableau_is_explicit(A: jax.typing.ArrayLike) -> jtp.Bool:
+        """
+        Check if the Butcher tableau corresponds to an explicit integration scheme.
+
+        Args:
+            A: The Runge-Kutta matrix.
+
+        Returns:
+            `True` if the Butcher tableau is explicit, `False` otherwise.
+        """
+
+        return jnp.allclose(A, jnp.tril(A, k=-1))
+
+    @staticmethod
+    def butcher_tableau_supports_fsal(
+        A: jax.typing.ArrayLike,
+        b: jax.typing.ArrayLike,
+        c: jax.typing.ArrayLike,
+        index_of_solution: jtp.IntLike = 0,
+    ) -> [bool, int | None]:
+        """
+        Check if the Butcher tableau supports the FSAL (first-same-as-last) property.
+
+        Args:
+            A: The Runge-Kutta matrix.
+            b: The weights coefficients.
+            c: The nodes coefficients.
+            index_of_solution:
+                The index of the row of `b.T` corresponding to the solution.
+
+        Returns:
+            A tuple containing a boolean indicating whether the Butcher tableau supports
+            FSAL, and the index i of the intermediate kᵢ derivative corresponding to the
+            initial derivative `f(x0, t0)` of the next step.
+        """
+
+        if not ExplicitRungeKutta.butcher_tableau_is_valid(A=A, b=b, c=c):
+            raise ValueError("The Butcher tableau is not valid.")
+
+        if not ExplicitRungeKutta.butcher_tableau_is_explicit(A=A):
+            return False
+
+        if index_of_solution >= b.T.shape[0]:
+            msg = "The index of the solution (i-th row of `b.T`) is out of range."
+            raise ValueError(msg)
+
+        if c[0] != 0:
+            return False, None
+
+        # Find all the rows of A where c = 1 (therefore at t=tf). The Butcher tableau
+        # supports FSAL if any of these rows (there might be more rows with c=1) matches
+        # the rows of b.T corresponding to the next state (marked by `index_of_solution`).
+        # This last condition means that the last kᵢ derivative is computed at (tf, xf),
+        # that corresponds to the (t0, x0) pair of the next integration call.
+        rows_of_A_with_fsal = (A == b.T[None, index_of_solution]).all(axis=1)
+        rows_of_A_with_fsal = jnp.logical_and(rows_of_A_with_fsal, (c == 1))
+
+        # If there is no match, it means that the Butcher tableau does not support FSAL.
+        if not rows_of_A_with_fsal.any():
+            return False, None
+
+        # Return the index of the row of A providing the fsal derivative (that is the
+        # possibly intermediate kᵢ derivative).
+        # Note that if multiple rows match (it should not), we return the first match.
+        return True, int(jnp.where(rows_of_A_with_fsal == True)[0].tolist()[0])
```

Comment:
This might create issues as it is an implicit boolean conversion of a traced array

```suggestion
        return True, int(jnp.where(rows_of_A_with_fsal is True)[0].tolist()[0])
```

Reply:
Generally speaking this would be true, but this method is not meant to run in any jitted context. I believe that in that case, the logic would fail first since the output of the method is not constant (either `None` or `int`).

The integrators are meant to be pytrees without leafs (for the moment) that are built outside transformed functions. Note, however, that `Integrator.step` could be jitted and vmapped.